lib/api/core/models/repositories/repository.js

```diff
@@ -1,344 +1,421 @@
+   * @param {object} [options] - Optional options.
+   * @param {string} [collection]
+   * @returns {Promise}
+   */
+  load (id, options, collection = this.collection) {
+    if (collection === null) {
+      return Promise.reject(new InternalError('The collection must be specified.'));
+    }
 
-/**
- * Given an object id, delete it from the configured storage engines
- *
- * @param {string} id   The id of the object to delete
- * The opts optional parameters currently accepts only 1 option:
- *   key: if provided, removes the given key instead of the default one (<collection>/<id>)
- * @param {object} [opts={}] optional options for the current operation
- * @returns {Promise}
- */
-Repository.prototype.delete = function repositoryDelete (id, opts) {
-  var promises = [];
+    if (!this.cacheEngine) {
+      return this.loadOneFromDatabase(id, collection);
+    }
 
-  if (this.cacheEngine) {
-    promises.push(this.deleteFromCache(id, opts));
+    return this.loadFromCache(id, options, collection)
+      .then(object => {
+        if (object === null) {
+          if (this.databaseEngine === null) {
+            return null;
+          }
+
+          return this.loadOneFromDatabase(id, collection)
+            .then(objectFromDatabase => {
+              if (objectFromDatabase !== null) {
+                this.persistToCache(objectFromDatabase, collection);
+              }
+              return objectFromDatabase;
+            });
+        }
+        this.refreshCacheTTL(object, collection);
+        return object;
+      });
   }
-  if (this.databaseEngine) {
-    promises.push(this.deleteFromDatabase(id));
+
+  /**
+   * Persists the given object in the collection that is attached to the repository.
+   *
+   * @param {Profile|Role|User} object - The object to persist
+   * @param {object} [options] - The persistence options
+   * @param {string} [collection]
+   * @returns {Promise}
+   */
+  persistToDatabase (object, options = {}, collection = this.collection) {
+    const method = options.method || 'createOrReplace';
+
+    if (collection === null) {
+      return Promise.reject(new InternalError('The collection must be specified.'));
+    }
+
+    return this.databaseEngine[method](collection, object._id, this.serializeToDatabase(object));
   }
 
-  return Promise.all(promises);
-};
+  /**
+   * Given an object id, delete it from the configured storage engines
+   *
+   * @param {string} id - The id of the object to delete
+   * The options optional parameters currently accepts only 1 option:
+   *   key: if provided, removes the given key instead of the default one (<collection>/<id>)
+   * @param {object} [options] - optional options for the current operation
+   * @param {string} [collection]
+   * @returns {Promise}
+   */
+  delete (id, options = {}, collection = this.collection) {
+    const promises = [];
+
+    if (collection === null) {
+      return Promise.reject(new InternalError('The collection must be specified.'));
```

Comment:
It can not happen due to a bad request, it might come from an bad plugin implementation, will use this one instead.